### ISSUE TITLE
Comms layer revamp

### ISSUE THREAD

**Issue by openhardwarecoza** [LaserWeb/lw.comms#1]
![revamp comms layer](https://cloud.githubusercontent.com/assets/7695323/20335374/bd5716f8-abca-11e6-84f0-2b41e1121f6e.PNG)

So, as the work on LW4 starts forging ahead, the time has come to start seriously working on revamping our communications layer

**Comment by openhardwarecoza** [comment]
Notes:

# TinyG:
https://github.com/giseburt/TinyG-node   by @giseburt
and
https://github.com/cheton/cnc/tree/f6d0bba4767912a7a813f4694af03e47c8ba2a50/src/app/controllers/TinyG2 by @cheton


# Grbl: 
https://github.com/cheton/cnc/tree/f6d0bba4767912a7a813f4694af03e47c8ba2a50/src/app/controllers/Grbl 
and 
https://github.com/grbl/grbl/wiki/Interfacing-with-Grbl#streaming-protocol-character-counting-recommended-with-reservations
@chamnit - whats your recommendation:  ping pong or char count. Any devs you feel we can tag in to help get this done at last.  @cheton also did some work here 

# Smoothie
How Arthur would love me to drop wifi and usb as options and only do ethernet. Sadly not realistic so we need to solve some headaches

So far the best we have is https://github.com/LaserWeb/LaserWeb3/blob/master/server-smoothie.js
Couple shortcomings: 
- Queue stops when error is encountered - needs better error handling
- ping pong buffer is causing buffer underrun for some raster operations. Can we use a char counting buffer here too? @wolfmanjm

# Overall
The hardest part of supporting all the firmwares is th differences between them.  When I become world dictator I am putting you all in one room till you leave with ONE standard set of commands and responses!  
We need these "comms modules" to return a standard set of data back and forth between the frontend. 
I can't keep doing "if then else" in the frontend. Its ridicolous

**Comment by openhardwarecoza** [comment]
Sorry, accidentally assigned...

I need volunteers (:
@Cinezaster @cheton @lautr3k

**Comment by wolfmanjm** [comment]
No you cannot use char counting for smoothie over USB. You can stream gcode over the telnet port for smoothie over network, it is just streamed and allows TCP/IP to handle flow control. (No pingpong).
I think it is incorrect that the ping pong over USB causes underrun, this is simply incorrect and not the reason it underuns. It is simply you are feeding gcode slower than it is being consumed, no amount of hacking the comms layer can fix that fact. If the gcode is a set of tiny segments then the only solution is to send a compressed stream instead of gcode. We can work on what that protocol would look like.

**Comment by openhardwarecoza** [comment]
The telnet based stream will again, only work for ethernet. 
So still stuck on a USB solution. Unless https://github.com/LaserWeb/LaserWeb3/issues/171 is the way to go

**Comment by jorgerobles** [comment]
So, GRBL over Wifi will be achieved using an ESP8266 / ESP32 as the broker?
It's worth mentioning because also could overcome GRBL limitations (monitor display, even SD) <:unicorn:>And ESP could be even the next GRBL core processor</:unicorn:>

**Comment by cprezzi** [comment]
I could also step into the comms layer revamp and backend dev, if needed.

**Comment by openhardwarecoza** [comment]
@jorgerobles yes, the current Wifi bridge does NOTHING except bridging serial<->websocket.  So theres not reason it shouldnt be suitable to bridge to Grbl too

@cprezzi absolutely I would take all the help I can get.  Just heads up. Its a hard one (:

**Comment by cheton** [comment]
It looks pretty good. The communication layer in this figure looks very similar to the server architecture of the cnc project (https://github.com/cheton/cnc), except that only serial port connection over USB is supported so far by cnc. For streaming G-code to Grbl, I'd also recommend using the character-counting protocol rather than the send-response method to maximize Grbl's performance. TinyG also has a good flow control mechanism by monitoring the planner queue status.

Just let me know if there is anything I can help.

**Comment by chamnit** [comment]
@openhardwarecoza : @cheton is correct. For laser raster applications, I would use the character-counting protocol for Grbl. This is what UGS uses and why it tends to perform well.

**Comment by openhardwarecoza** [comment]
@cheton is your Implementation using the char counter already?  (: if you
have a couple hours to spare,  take a look at
https://github.com/LaserWeb/LaserWeb3/blob/master/server-grbl.js and also
https://github.com/LaserWeb/LaserWeb3/blob/master/server-smoothie.js and
shoot here with thoughts on how it can be improved.  Our cam side is
strong,  but comms side has always been a little less than perfect

**Comment by cheton** [comment]
So far I'm still using the send-response method, but implementing character-counting protocol is already in the roadmap. I will make it ready in a future version.

**Comment by giseburt** [comment]
@cheton, using the TinyG module (soon to become the g2core-api module), you don't have to worry about flow control at all. We don't recommend using the QR in any case.

We are also working on a REST+websocket server that's also written in node, and sits on top of the g2core-api module.

Let me know if you have any questions. I don't have time to help write much, but I'm certainly available for questions.
